spec/features/admin/products_spec.rb

```diff
@@ -114,6 +114,71 @@
       expect(page).to have_content "Unit value can't be blank"
     end
   end
+  
+  describe "deleting", js: true do
+    let!(:product1) { create(:simple_product, name: 'a product to keep', supplier: @supplier) }
+
+    context 'a simple product' do
+      let!(:product2) { create(:simple_product, name: 'a product to delete', supplier: @supplier) }
+
+      before do
+        login_as_admin_and_visit spree.admin_products_path
+
+        within "tr#p_#{product2.id}" do
+          accept_alert { page.find("[data-powertip=Remove]").click }
+        end
+      end
+
+      it 'removes it from the product list' do
+        expect(page).not_to have_selector "tr#p_#{product2.id}"
+        expect(page).to have_selector "tr#p_#{product1.id}"
+      end
+    end
+
+    context 'a shipped product' do
+      let!(:order) { create(:shipped_order, line_items_count: 1) }
+      let!(:line_item) { order.reload.line_items.first }
+
+      before do
+        login_as_admin_and_visit spree.admin_products_path
+
+        within "tr#p_#{order.variants.first.product_id}" do
+          accept_alert { page.find("[data-powertip=Remove]").click }
+        end
+      end
+
+      it 'removes it from the product list' do
+        expect(page).to have_selector "tr#p_#{product1.id}"
+        expect(page).not_to have_selector "tr#p_#{order.variants.first.product_id}"
+        expect(Spree::Product.count).to eq 1
+        expect(order.line_items.count).to eq 1
```

Comment:
Likewise, I found it also leads to more robust feature specs to rely on UI only at this testing-level. The idea is that everything is checked counting UI elements, or navigating to other index pages that do show the information we need. That's what we do when testing manually and is less dependant on the implementation underneath. Therefore, it gives us more confidence while refactoring.

Reply:
That's a good idea; I removed those two lines as the result might be misleading. As you say, the idea here is to test the UI.

I felt it was a good idea to add a `visit current_path` in the respective `before` block. I think the assertion has value only after page reload: this way we are sure that the DB contents are being used to populate the page.